File: livingrimoire start here/LivinGrimoire Python annotationless version/livingrimoire.py
from collections import deque  # for APVerbatim cls


class AbsDictionaryDB:
    def save(self, key, value):
        pass

    # noinspection PyMethodMayBeStatic
    def load(self, key):
        return "null"


class AlgPart:
    # one part of an algorithm, it is a basic simple action or sub goal
    def __init__(self):
        # set True to stop the entire running active Algorithm
        self.algKillSwitch = False

    def action(self, ear, skin, eye):
        """Returns action string"""
        pass

    def completed(self):
        """Has finished ?"""
        pass

    def myName(self):
        """Returns the class name"""
        return self.__class__.__name__


class APVerbatim(AlgPart):
    def __init__(self, *sentences):
        super().__init__()
        self.sentences = deque(sentences)
        # Handle the case where a single list is passed (like the ArrayList constructor in Java)
        if len(sentences) == 1 and isinstance(sentences[0], list):
            self.sentences = deque(sentences[0])

    def action(self, ear, skin, eye):
        # Use deque.popleft() safely without try-except
        return self.sentences.popleft() if self.sentences else ""

    def completed(self):
        return not self.sentences


# A step-by-step plan to achieve a goal
class Algorithm:

    def __init__(self, algParts):  # list of Mutatable
        super().__init__()
        self.algParts = algParts

    @classmethod
    def from_varargs(cls, *algParts):
        # Create an instance from varargs
        return cls(list(algParts))

    @property
    def getAlgParts(self):
        return self.algParts

    def getSize(self):
        return len(self.algParts)


# the Kokoro class enables: using a database, inter skill communication and action log monitoring
class Kokoro:
    def __init__(self, absDictionaryDB):
        self.emot = ""
        self.grimoireMemento = absDictionaryDB
        self.toHeart = {}

    def getEmot(self):
        return self.emot

    def setEmot(self, emot):
        self.emot = emot


# used to transport algorithms
class Neuron:
    def __init__(self):
        self._defcons = {}
        for i in range(1, 6):
            self._defcons[i] = []

    def insertAlg(self, priority, alg):
        if 0 < priority < 6:
            if len(self._defcons[priority]) < 4:
                self._defcons[priority].append(alg)

    def getAlg(self, defcon):
        if len(self._defcons[defcon]) > 0:
            temp = self._defcons[defcon].pop(0)
            return temp
        return None


class Skill:
    def __init__(self):
        # The variables start with an underscore (_) because they are protected
        self._kokoro = None  # consciousness, shallow ref for interskill comms
        self._outAlg = None  # skills output
        self._outpAlgPriority = -1  # defcon 1->5
        self._skill_type = 1  # 1:regular, 2:aware_skill, 3:continuous_skill
        self._skill_lobe = 1  # 1:logical, 2:hardware, 3:ear, 4:skin, 5:eye Chobits

    def setOutalg(self, alg):
        self._outAlg = alg

    def getOutAlg(self):
        return self._outAlg

    def setOutAlgPriority(self, priority):
        self._outpAlgPriority = priority

    # skill triggers and algorithmic logic
    def input(self, ear, skin, eye):
        pass

    # extraction of skill algorithm to run (if there is one)
    def output(self, neuron):
        if self._outAlg is not None:
            neuron.insertAlg(self._outpAlgPriority, self._outAlg)
            self._outpAlgPriority = -1
            self._outAlg = None

    def setKokoro(self, kokoro):
        # telepathic communication between chobits
        self._kokoro = kokoro

    def getKokoro(self):
        return self._kokoro

    # --- Algorithm shortcuts ---
    def setVerbatimAlg(self, priority, *sayThis):
        # build a simple speak-by-string algorithm
        self._outAlg = Algorithm.from_varargs(APVerbatim(*sayThis))
        self._outpAlgPriority = priority  # 1=highest priority

    def setSimpleAlg(self, *sayThis):
        # shortcut for low-priority output
        self._outAlg = Algorithm.from_varargs(APVerbatim(*sayThis))
        self._outpAlgPriority = 4  # default low priority

    def setVebatimAlgFromList(self, priority, sayThis):
        # list version of verbatim
        self._outAlg = Algorithm.from_varargs(APVerbatim(sayThis))
        self._outpAlgPriority = priority

    def algPartsFusion(self, priority, *algParts):
        # fuse arbitrary AlgParts into an algorithm
        self._outAlg = Algorithm.from_varargs(*algParts)
        self._outpAlgPriority = priority

    def pendingAlgorithm(self):
        # is an algorithm pending?
        return self._outAlg is not None

    def skillNotes(self, param):
        return "notes unknown"

    # Getter and Setter for skill_type
    def get_skill_type(self):
        return self._skill_type

    def set_skill_type(self, skill_type):
        if 1 <= skill_type <= 3:
            self._skill_type = skill_type

    # Getter and Setter for skill_lobe
    def get_skill_lobe(self):
        return self._skill_lobe

    def set_skill_lobe(self, skill_lobe):
        if 1 <= skill_lobe <= 5:
            self._skill_lobe = skill_lobe


class DiHelloWorld(Skill):
    # Override
    def input(self, ear, skin, eye):
        if ear == "hello":
            self.setVerbatimAlg(4, "hello world")  # 1->5 (1 = highest priority)

    def skillNotes(self, param):
        if param == "notes":
            return "plain hello world skill"
        elif param == "triggers":
            return "say hello"
        return "note unavalible"


class Cerebellum:
    # runs an algorithm
    def __init__(self):
        self.fin = None
        self.at = None
        self.incrementAt = False
        self.alg = None
        self.isActive = False
        self.emot = ""

    def advanceInAlg(self):
        if self.incrementAt:
            self.incrementAt = False
            self.at += 1
            if self.at == self.fin:
                self.isActive = False

    def getAt(self):
        return self.at

    def getEmot(self):
        return self.emot

    def setAlgorithm(self, algorithm):
        if not self.isActive and (algorithm.getAlgParts is not None):
            self.alg = algorithm
            self.at = 0
            self.fin = algorithm.getSize()
            self.isActive = True
            self.emot = self.alg.getAlgParts[self.at].myName()  # updated line

    def isActiveMethod(self):
        return self.isActive

    def act(self, ear, skin, eye):
        axnStr = ""
        if not self.isActive:
            return axnStr
        if self.at < self.fin:
            axnStr = self.alg.getAlgParts[self.at].action(ear, skin, eye)
            self.emot = self.alg.getAlgParts[self.at].myName()
            if self.alg.getAlgParts[self.at].completed():
                self.incrementAt = True
        return axnStr

    def deActivateAlg(self):
        # stop the entire running active Algorithm
        self.isActive = self.isActive and not self.alg.getAlgParts[self.at].algKillSwitch


class Fusion:
    def __init__(self):
        self._emot = ""
        self.ceraArr = [Cerebellum() for _ in range(5)]
        self._result = ""

    def getEmot(self):
        return self._emot

    def loadAlgs(self, neuron):
        for i in range(1, 6):
            if not self.ceraArr[i - 1].isActive:
                temp = neuron.getAlg(i)
                if temp is not None:
                    self.ceraArr[i - 1].setAlgorithm(temp)

    def runAlgs(self, ear, skin, eye):
        self._result = ""
        for i in range(5):
            if not self.ceraArr[i].isActive:
                continue
            self._result = self.ceraArr[i].act(ear, skin, eye)
            self.ceraArr[i].advanceInAlg()
            self._emot = self.ceraArr[i].getEmot()
            self.ceraArr[i].deActivateAlg()  # deactivation if AlgPart.algkillswitch = true
            return self._result
        self._emot = ""
        return self._result


class Chobits:
    def __init__(self):
        super().__init__()
        self.dClasses = []  # regular skills
        self._fusion = Fusion()
        self._neuron = Neuron()
        self._kokoro = Kokoro(AbsDictionaryDB())  # soul
        self._isThinking = False
        self._awareSkills = []  # self awareness skills. Chobit Object ref attribute.
        self.alg_triggered = False
        self.cts_skills = []

    def setDatabase(self, absDictionaryDB):
        self._kokoro.grimoireMemento = absDictionaryDB

    def add_regular_skill(self, skill):
        # add a skill (builder design patterned func))
        if self._isThinking:
            return self
        skill.set_skill_type(1)
        skill.setKokoro(self._kokoro)
        self.dClasses.append(skill)
        return self

    def addSkillAware(self, skill):
        # add a skill with Chobit Object in their c'tor
        skill.set_skill_type(2)
        skill.setKokoro(self._kokoro)
        self._awareSkills.append(skill)

    def add_continuous_skill(self, skill):
        # recommended for skills that trigger non stop or even burst mode triggered.
        if self._isThinking:
            return
        skill.set_skill_type(3)
        skill.setKokoro(self._kokoro)
        self.cts_skills.append(skill)

    def clear_regular_skills(self):
        # remove all logical(regular) skills
        if self._isThinking:
            return
        self.dClasses.clear()

    def clear_continuous_skills(self):
        if self._isThinking:
            return
        self.cts_skills.clear()

    def clear_all_skills(self):
        self.clear_regular_skills()
        self.clear_continuous_skills()

    def addSkills(self, *skills):
        if self._isThinking:
            return
        for skill in skills:
            skill.setKokoro(self._kokoro)
            self.dClasses.append(skill)

    def remove_logical_skill(self, skill):
        if self._isThinking:
            return
        if skill not in self.dClasses:
            return
        self.dClasses.remove(skill)

    def remove_continuous_skill(self, skill):
        if self._isThinking:
            return
        if skill not in self.cts_skills:
            return
        self.cts_skills.remove(skill)

    def remove_skill(self, skill):
        """remove any type of skill (except aware skills)"""
        if skill.get_skill_type() == 1:
            self.remove_logical_skill(skill)
        else:
            self.remove_continuous_skill(skill)

    def containsSkill(self, skill):
        return skill in self.dClasses

    def think(self, ear, skin, eye):
        self.alg_triggered = False
        # main skill loop
        self._isThinking = True
        for dCls in self.dClasses:
            self.inOut(dCls, ear, skin, eye)
        self._isThinking = False
        # loop for skills with access to the Chobit Object:
        for dCls2 in self._awareSkills:
            self.inOut(dCls2, ear, skin, eye)
        # continuous skills loop
        self._isThinking = True
        for d_cls3 in self.cts_skills:
            if self.alg_triggered:
                break
            self.inOut(d_cls3, ear, skin, eye)
        self._isThinking = False
        self._fusion.loadAlgs(self._neuron)
        return self._fusion.runAlgs(ear, skin, eye)

    def getSoulEmotion(self):
        # get AlgPart class name representing emotion
        return self._fusion.getEmot()

    def inOut(self, dClass, ear, skin, eye):
        dClass.input(ear, skin, eye)  # new
        if dClass.pendingAlgorithm():
            self.alg_triggered = True
        dClass.output(self._neuron)

    def getKokoro(self):
        # several chobits can use the same soul
        return self._kokoro

    def setKokoro(self, kokoro):
        # use this for telepathic communication between different chobits objects
        self._kokoro = kokoro

    def get_skill_list(self):
        # get skill list of names(str)
        result = []
        for skill in self.dClasses:
            result.append(skill.__class__.__name__)
        return result

    def add_skill(self, skill):
        """
        Automatically adds a skill to the correct category based on its type.
        No manual classification needed—just pass the skill and let the system handle it.
        """
        match skill.get_skill_type():
            case 1:  # Regular Skill
                self.add_regular_skill(skill)
            case 2:  # Aware Skill
                self.addSkillAware(skill)
            case 3:  # Continuous Skill
                self.add_continuous_skill(skill)


class Brain:
    # c'tor
    def __init__(self):
        self._emotion = ""
        self._logicChobitOutput = ""
        self.logicChobit = Chobits()
        self.hardwareChobit = Chobits()
        #120425 upgrade
        self.ear = Chobits()
        self.skin = Chobits()
        self.eye = Chobits()
        Brain.__imprintSoul(self.logicChobit.getKokoro(), self.hardwareChobit, self.ear, self.skin, self.eye)

    @staticmethod
    def __imprintSoul(kokoro, *args):
        for arg in args:
            arg.setKokoro(kokoro)

    # ret active alg part representing emotion
    def getEmotion(self):
        return self._emotion

    # ret feedback (last output)
    def getLogicChobitOutput(self):
        return self._logicChobitOutput

    # live
    def doIt(self, ear, skin, eye):
        self._logicChobitOutput = self.logicChobit.think(ear, skin, eye)
        self._emotion = self.logicChobit.getSoulEmotion()
        self.hardwareChobit.think(self._logicChobitOutput, skin, eye)

    def add_skill(self, skill):
        """
        Adds a skill to the correct Chobits based on its skill_lobe attribute.
        Just pass the skill—the Brain handles where it belongs.
        """
        match skill.get_skill_lobe():
            case 1:  # Logical skill
                self.logicChobit.add_skill(skill)
            case 2:  # Hardware skill
                self.hardwareChobit.add_skill(skill)
            case 3:  # Ear skill
                self.ear.add_skill(skill)
            case 4:  # Skin skill
                self.skin.add_skill(skill)
            case 5:  # Eye skill
                self.eye.add_skill(skill)

    # add regular thinking(logical) skill
    def add_logical_skill(self, skill):
        skill.set_skill_lobe(1)
        self.logicChobit.add_regular_skill(skill)

    # add output skill
    def add_hardware_skill(self, skill):
        skill.set_skill_lobe(2)
        self.hardwareChobit.add_regular_skill(skill)

    def add_skillAware(self, skill):
        # add skill with a Chobits reference attribute.
        skill.set_skill_lobe(1)
        self.logicChobit.addSkillAware(skill)

    # add audio(ear) input skill
    def add_ear_skill(self, skill):
        skill.set_skill_lobe(3)
        self.ear.add_regular_skill(skill)

    # add sensor input skill
    def add_skin_skill(self, skill):
        skill.set_skill_lobe(4)
        self.skin.add_regular_skill(skill)

    # add visual input skill
    def add_eye_skill(self, skill):
        skill.set_skill_lobe(5)
        self.eye.add_regular_skill(skill)

    def think_default(self, keyIn):
        """
        The primary thinking method—automatically processes input whether it's
        typed or received through sensory channels.
        """
        if bool(keyIn):
            # handles typed inputs(keyIn)
            self.doIt(keyIn, "", "")  # the string is not empty
        else:
            # the string is empty, process with sensory inputs
            self.doIt(self.ear.think("", "", ""), self.skin.think("", "", ""), self.eye.think("", "", ""))

    def think(self):
        # overload of think method because Python does not support it
        self.doIt(self.ear.think("", "", ""), self.skin.think("", "", ""), self.eye.think("", "", ""))


class DiSysOut(Skill):
    def __init__(self):
        super().__init__()
        self._skill_lobe = 2  # hardware chobit skill

    def input(self, ear, skin, eye):
        if ear and "#" not in ear:
            print(ear)  # raw interdimensional stdout!

    def skillNotes(self, param):
        if param == "notes":
            return "prints to console"
        elif param == "triggers":
            return "automatic for any input"
        return "note unavalible"

File: livingrimoire start here/LivinGrimoire swift/livingrimoire packet/LivinGrimoire.swift
//
//  LivinGrimoire.swift
//  LivinGrimoireSwiftV1
//
//  Created by moti barski on 21/06/2022.
//

import Foundation

open class AbsDictionaryDB{
    func save(key:String, value: String){
        // save to DB (override me)
    }
    func load(key:String)->String{
        // override me
        return "null"
    }
}

open class AlgPart {
    var algKillSwitch: Bool = false
    
    func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        return ""
    }
    
    func completed() -> Bool {
        // Has finished?
        return false
    }
    
    func myName() -> String {
        // Returns the class name
        return String(describing: type(of: self))
    }
}


class APVerbatim: AlgPart {
    private var buffer: [String]
    private var headIndex: Int = 0
    
    init(_ sentences: String...) {
        self.buffer = sentences
    }
    
    init(_ list1: [String]) {
        self.buffer = list1
    }
    
    override func action(_ ear: String, _ skin: String, _ eye: String) -> String {
        guard headIndex < buffer.count else { return "" }
        defer { headIndex += 1 }
        return buffer[headIndex]
    }
    
    override func completed() -> Bool {
        headIndex >= buffer.count
    }
    
    func compact() {
        if headIndex > 1024 {
            buffer.removeFirst(headIndex)
            headIndex = 0
        }
    }
}

// A step-by-step plan to achieve a goal
class Algorithm {
    private var algParts: [AlgPart]
    
    init(_ algParts: [AlgPart]) {
        self.algParts = algParts
    }
    
    init(_ algParts: AlgPart...) {
        self.algParts = algParts
    }
    
    func getAlgParts() -> [AlgPart] {
        return algParts
    }
    
    func getSize() -> Int {
        return algParts.count
    }
}

/* This class enables:
   - Communication between skills
   - Utilization of a database for skills
   This class is a built-in attribute in skill objects.
*/
class Kokoro {
    private var emot: String = ""
    var grimoireMemento: AbsDictionaryDB
    var toHeart: [String: String] = [:]
    
    init(_ absDictionaryDB: AbsDictionaryDB) {
        self.grimoireMemento = absDictionaryDB
    }
    
    func getEmot() -> String {
        return emot
    }
    
    func setEmot(_ emot: String) {
        self.emot = emot
    }
}

// Used to transport algorithms to other classes
class Neuron {
    private var defcons: [Int: [Algorithm]] = [:]
    
    init() {
        for i in 1..<6 {
            defcons[i] = []
        }
    }
    
    func insertAlg(_ priority: Int, _ alg: Algorithm) {
        guard (0 < priority) && (priority < 6) else { return }
        guard var algs = defcons[priority], algs.count < 4 else { return }
        algs.append(alg)
        defcons[priority] = algs
    }
    
    func getAlg(_ defcon: Int) -> Algorithm? {
        guard var algs = defcons[defcon], !algs.isEmpty else { return nil }
        let temp = algs.removeFirst()
        defcons[defcon] = algs
        return temp
    }
}

// Skill: Base class for handling AI-driven skills
open class Skill {
    
    var kokoro: Kokoro? = nil // Consciousness, enables interskill communication
    var outAlg: Algorithm? = nil // Skill output algorithm
    var outpAlgPriority: Int = -1 // Defcon 1->5
    
    fileprivate var skill_type: Int = 1 // 1: Regular, 2: Aware Skill, 3: Continuous Skill
    fileprivate var skill_lobe: Int = 1 // 1: Logical, 2: Hardware, 3: Ear, 4: Skin, 5: Eye Chobits
    
    init() {}
    
    // Skill triggers and algorithmic logic
    func input(_ ear: String, _ skin: String, _ eye: String) {}
    
    // Extracts skill algorithm to run (if one exists)
    func output(_ neuron: Neuron) {
        if let alg = outAlg {
            neuron.insertAlg(outpAlgPriority, alg)
            outpAlgPriority = -1
            outAlg = nil
        }
    }
    
    func setKokoro(_ kokoro: Kokoro) {
        // Use for telepathic communication between Chobits objects
        self.kokoro = kokoro
    }
    
    // Algorithm building shortcut methods
    func setVerbatimAlg(_ priority: Int, _ sayThis: String...) {
        // Build a simple output algorithm to speak string-by-string per think cycle
        self.outAlg = Algorithm(APVerbatim(sayThis))
        self.outpAlgPriority = priority // 1->5, 1 is the highest priority
    }
    
    func setSimpleAlg(_ sayThis: String...) {
        // Similar to setVerbatimAlg, defaulting to priority 4
        self.outAlg = Algorithm(APVerbatim(sayThis))
        self.outpAlgPriority = 4
    }
    
    func setVerbatimAlgFromList(_ priority: Int, _ sayThis: [String]) {
        // Build a simple output algorithm using a list parameter
        self.outAlg = Algorithm(APVerbatim(sayThis))
        self.outpAlgPriority = priority
    }
    
    func algPartsFusion(_ priority: Int, _ algParts: AlgPart...) {
        self.outAlg = Algorithm(algParts)
        self.outpAlgPriority = priority
    }
    
    func pendingAlgorithm() -> Bool {
        // Is an algorithm pending?
        return outAlg != nil
    }
    
    // Getter and Setter for skill_type
    func getSkillType() -> Int {
        return skill_type
    }
    
    func setSkillType(_ skill_type: Int) {
        // 1: Regular, 2: Aware Skill, 3: Continuous Skill
        if (1...3).contains(skill_type) {
            self.skill_type = skill_type
        }
    }
    
    // Getter and Setter for skill_lobe
    func getSkillLobe() -> Int {
        return skill_lobe
    }
    
    func setSkillLobe(_ skill_lobe: Int) {
        // 1: Logical, 2: Hardware, 3: Ear, 4: Skin, 5: Eye Chobits
        if (1...5).contains(skill_lobe) {
            self.skill_lobe = skill_lobe
        }
    }
    
    func skillNotes(_ param: String) -> String {
        return "notes unknown"
    }
}


class DiHelloWorld: Skill {
    // hello world skill for testing purposes
    override init() {
        super.init()
    }

    override func input(_ ear: String, _ skin: String, _ eye: String) {
        switch ear {
        case "hello":
            super.setSimpleAlg("hello world") // 1->5 1 is the highest algorithm priority
        default:
            break
        }
    }

    override func skillNotes(_ param: String) -> String {
        if param == "notes" {
            return "plain hello world skill"
        } else if param == "triggers" {
            return "say hello"
        }
        return "note unavailable"
    }
}

class Cerebellum {
    // Runs an algorithm
    private var fin: Int = 0
    private var at: Int = 0
    private var incrementAt: Bool = false
    
    var alg: Algorithm?
    private var isActive: Bool = false
    private var emot: String = ""
    
    func advanceInAlg() {
        if incrementAt {
            incrementAt = false
            at += 1
            if at == fin {
                isActive = false
            }
        }
    }
    
    func getAt() -> Int {
        return at
    }
    
    func getEmot() -> String {
        return emot
    }
    
    func setAlgorithm(_ algorithm: Algorithm) {
        if !isActive && !algorithm.getAlgParts().isEmpty {
            self.alg = algorithm
            self.at = 0
            self.fin = algorithm.getSize()
            self.isActive = true
            self.emot = algorithm.getAlgParts()[at].myName()
        }
    }
    
    func getIsActive() -> Bool {
        return isActive
    }
    
    func act(_ ear: String, _ skin: String, _ eye: String) -> String {
        var axnStr = ""
        guard isActive else { return axnStr }
        
        if at < fin, let currentAlg = alg {
            axnStr = currentAlg.getAlgParts()[at].action(ear, skin, eye)
            self.emot = currentAlg.getAlgParts()[at].myName()
            
            if currentAlg.getAlgParts()[at].completed() {
                incrementAt = true
            }
        }
        return axnStr
    }
    
    func deactivate() {
        if isActive, let currentAlg = alg {
            isActive = !currentAlg.getAlgParts()[at].algKillSwitch
        }
    }
}

class Fusion {
    private var emot: String = ""
    private var result: String = ""
    private var ceraArr: [Cerebellum] = Array(repeating: Cerebellum(), count: 5)
    
    public init() {
        for i in 0..<5 {
            ceraArr[i] = Cerebellum()
        }
    }
    
    func getEmot() -> String {
        return emot
    }
    
    func loadAlgs(_ neuron: Neuron) {
        for i in 1...5 {
            if !ceraArr[i-1].getIsActive() {
                if let temp = neuron.getAlg(i) {
                    ceraArr[i-1].setAlgorithm(temp)
                }
            }
        }
    }
    
    func runAlgs(_ ear: String, _ skin: String, _ eye: String) -> String {
        result = ""
        for i in 0..<5 {
            guard ceraArr[i].getIsActive() else { continue }
            
            result = ceraArr[i].act(ear, skin, eye)
            ceraArr[i].advanceInAlg()
            emot = ceraArr[i].getEmot()
            ceraArr[i].deactivate() // deactivation if Mutatable.algkillswitch = true
            return result
        }
        emot = ""
        return result
    }
}

// The Chobits class represents an AI entity managing skills and algorithms
open class Chobits {
    
    public var dClasses: [Skill] = [] // Regular skills
    fileprivate var fusion: Fusion
    fileprivate var neuron: Neuron
    fileprivate var kokoro: Kokoro // consciousness
    
    private var isThinking: Bool = false
    private var awareSkills: [Skill] = []
    public var algTriggered: Bool = false
    public var cts_skills: [Skill] = [] // Continuous skills
    
    public init() {
        // Constructor
        self.fusion = Fusion()
        self.neuron = Neuron()
        self.kokoro = Kokoro(AbsDictionaryDB()) // consciousness initialization
    }
    
    func setDatabase(_ absDictionaryDB: AbsDictionaryDB) {
        self.kokoro.grimoireMemento = absDictionaryDB
    }
    
    func addRegularSkill(_ skill: Skill) {
        // Add a skill (builder design patterned func)
        guard !isThinking else { return }
        skill.setSkillType(1)
        skill.setKokoro(kokoro)
        dClasses.append(skill)
    }
    
    func addSkillAware(_ skill: Skill) {
        // Add a skill with Chobit Object in their constructor
        skill.setSkillType(2)
        skill.setKokoro(kokoro)
        awareSkills.append(skill)
    }
    
    func addContinuousSkill(_ skill: Skill) {
        // Add a skill (builder design patterned func)
        guard !isThinking else { return }
        skill.setSkillType(3)
        skill.setKokoro(kokoro)
        cts_skills.append(skill)
    }
    
    func clearRegularSkills() {
        // Remove all regular skills
        guard !isThinking else { return }
        dClasses.removeAll()
    }
    
    func clearContinuousSkills() {
        // Remove all continuous skills
        guard !isThinking else { return }
        cts_skills.removeAll()
    }
    
    func clearAllSkills() {
        clearRegularSkills()
        clearContinuousSkills()
    }
    
    func addSkills(_ skills: Skill...) {
        for skill in skills {
            addSkill(skill)
        }
    }
    
    func removeLogicalSkill(_ skill: Skill) {
        guard !isThinking else { return }
        dClasses.removeAll { $0 === skill }
    }
    
    func removeContinuousSkill(_ skill: Skill) {
        guard !isThinking else { return }
        cts_skills.removeAll { $0 === skill }
    }
    
    func removeSkill(_ skill: Skill) {
        /* Remove any type of skill (except aware skills) */
        if skill.getSkillType() == 1 {
            removeLogicalSkill(skill)
        } else {
            removeContinuousSkill(skill)
        }
    }
    
    func containsSkill(_ skill: Skill) -> Bool {
        return dClasses.contains { $0 === skill }
    }
    
    @discardableResult
    func think(_ ear: String, _ skin: String, _ eye: String) -> String {
        algTriggered = false
        isThinking = true // Regular skills loop
        
        for dCls in dClasses {
            inOut(dCls, ear, skin, eye)
        }
        
        isThinking = false
        
        for dCls2 in awareSkills {
            inOut(dCls2, ear, skin, eye)
        }
        
        isThinking = true
        
        for dCls2 in cts_skills {
            if algTriggered { break }
            inOut(dCls2, ear, skin, eye)
        }
        
        isThinking = false
        
        fusion.loadAlgs(neuron)
        return fusion.runAlgs(ear, skin, eye)
    }
    
    func getSoulEmotion() -> String {
        // Get the last active AlgPart name
        // The AP is an action, and it also represents an emotion
        return fusion.getEmot()
    }
    
    func inOut(_ dClass: Skill, _ ear: String, _ skin: String, _ eye: String) {
        dClass.input(ear, skin, eye) // Process input
        
        if dClass.pendingAlgorithm() {
            algTriggered = true
        }
        
        dClass.output(neuron)
    }
    
    func getKokoro() -> Kokoro {
        // Several Chobits can use the same soul
        // This enables telepathic communications between Chobits in the same project
        return kokoro
    }
    
    func setKokoro(_ kokoro: Kokoro) {
        // Use this for telepathic communication between different Chobits objects
        self.kokoro = kokoro
    }
    
    func getSkillList() -> [String] {
        return dClasses.map { String(describing: type(of: $0)) }
    }
    
    func getFusedSkills() -> [Skill] {
        /*
         Returns a fusion list containing both dClasses (regular skills)
         and cts_skills (continuous skills).
         */
        return dClasses + cts_skills
    }
    
    func addSkill(_ skill: Skill) {
        /*
         Automatically adds a skill to the correct category based on its type.
         No manual classification needed—just pass the skill and let the system handle it.
         */
        switch skill.getSkillType() {
        case 1:  // Regular Skill
            addRegularSkill(skill)
        case 2:  // Aware Skill
            addSkillAware(skill)
        case 3:  // Continuous Skill
            addContinuousSkill(skill)
        default:
            break
        }
    }
}

// The Brain class represents an AI core managing multiple Chobits
open class Brain {
    
    public var logicChobit: Chobits = Chobits()
    private var emotion: String = ""
    private var logicChobitOutput: String = ""
    
    public var hardwareChobit: Chobits = Chobits()
    public var ear: Chobits = Chobits()
    public var skin: Chobits = Chobits()
    public var eye: Chobits = Chobits()
    
    // Returns active algorithm part representing emotion
    func getEmotion() -> String {
        return emotion
    }
    
    // Returns feedback (last output)
    func getLogicChobitOutput() -> String {
        return logicChobitOutput
    }
    
    // Constructor
    init() {
        Brain.imprintSoul(logicChobit.getKokoro(), hardwareChobit, ear, skin, eye)
    }
    
    private static func imprintSoul(_ kokoro: Kokoro, _ args: Chobits...) {
        for arg in args {
            arg.setKokoro(kokoro)
        }
    }
    
    // Live processing method
    func doIt(_ ear: String, _ skin: String, _ eye: String) {
        logicChobitOutput = logicChobit.think(ear, skin, eye)
        emotion = logicChobit.getSoulEmotion()
        hardwareChobit.think(logicChobitOutput, skin, eye)
    }
    
    func addSkill(_ skill: Skill) {
        /*
         Adds a skill to the correct Chobits based on its skill_lobe attribute.
         Just pass the skill—the Brain handles where it belongs.
         */
        switch skill.getSkillLobe() {
        case 1:  // Logical skill
            logicChobit.addSkill(skill)
        case 2:  // Hardware skill
            hardwareChobit.addSkill(skill)
        case 3:  // Ear skill
            ear.addSkill(skill)
        case 4:  // Skin skill
            skin.addSkill(skill)
        case 5:  // Eye skill
            eye.addSkill(skill)
        default:
            break
        }
    }
    
    func chained(_ skill: Skill) -> Brain {
        // Chained add skill
        addSkill(skill)
        return self
    }
    
    // Add logical processing skill
    func addLogicalSkill(_ skill: Skill) { logicChobit.addRegularSkill(skill) }
    
    // Add hardware output skill
    func addHardwareSkill(_ skill: Skill) { hardwareChobit.addRegularSkill(skill) }
    
    // Add audio (ear) input skill
    func addEarSkill(_ skill: Skill) { ear.addRegularSkill(skill) }
    
    // Add sensor input skill
    func addSkinSkill(_ skill: Skill) { skin.addRegularSkill(skill) }
    
    // Add visual input skill
    func addEyeSkill(_ skill: Skill) { eye.addRegularSkill(skill) }
    
    func think(_ keyIn: String) {
        if !keyIn.isEmpty {
            // Handles typed inputs (keyIn)
            doIt(keyIn, "", "")
        } else {
            // Accounts for sensory inputs
            doIt(ear.think("", "", ""), skin.think("", "", ""), eye.think("", "", ""))
        }
    }
    
    func think() {
        // Accounts for sensory inputs only. Use this overload for tick events (where no typed inputs are expected)
        doIt(ear.think("", "", ""), skin.think("", "", ""), eye.think("", "", ""))
    }
}

// DiSysOut: Continuous skill that prints ear input to the console
class DiSysOut: Skill {
    
    override init() {
        super.init()
        setSkillType(3) // Continuous skill
        setSkillLobe(2) // Hardware Chobit
    }
    
    override func input(_ ear: String, _ skin: String, _ eye: String) {
        if !ear.isEmpty && !ear.contains("#") {
            print(ear)
        }
    }
    
    override func skillNotes(_ param: String) -> String {
        switch param {
        case "notes":
            return "prints to console"
        case "triggers":
            return "automatic for any input"
        default:
            return "note unavailable"
        }
    }
}

